### PR TITLE
Adding iree_atomic_slist_t type.

### PR DESCRIPTION
**Pull request by benvanik**
Well, a _somewhat atomic_ slist anyway :)
What matters today is that this _can_ be golfed down to single atomic ops, but this is blocking progress on threading work so I am playing it safe here.

### CODE REVIEW

File: iree/base/atomic_slist.c
Comment (ScottTodd) [inline iree/base/atomic_slist.c:21]: 404? https://reviews.llvm.org/D18500 works
```suggestion
// https://reviews.llvm.org/D18500
```

File: iree/base/atomic_slist.h
Comment (ScottTodd) [inline iree/base/atomic_slist.h:189]: Neat, these macros work basically like a C++ templated class but in C?

**Review by ScottTodd — APPROVED** [review]
Wait until `benvanik-hack-and-slash` is merged, then rebase onto `main`?

File: iree/base/atomic_slist.h
Comment (benvanik) [reply iree/base/atomic_slist.h:189]: yep! so you'll get nice type errors if you try to use the wrong types together and such.

**Pull request merged**
